File: notifier/src/main/java/com/event/processing/notifier/domain/repository/SegmentRepository.java
package com.event.processing.notifier.domain.repository;

import com.event.processing.notifier.domain.dto.SegmentDTO;
import com.event.processing.notifier.domain.entity.Segment;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Set;

/**
 * Repository interface for managing Segment entities.
 * Provides database operations for segments and their relationships with
 * subscribers.
 * <p>
 * Key features:
 * - Basic CRUD operations inherited from JpaRepository
 * - Custom query for fetching segments by subscriber IDs
 * - DTO-based segment retrieval
 *
 * @author LongLe
 * @version 1.0
 */
@Repository
public interface SegmentRepository extends JpaRepository<Segment, String> {
  /**
   * Fetches segments associated with the given subscriber IDs.
   * Uses a JPQL query to join the Segment and SubscriberSegment tables
   * and returns the results as SegmentDTO objects.
   *
   * @param subscriberIds Set of subscriber IDs to fetch segments for
   * @return List of SegmentDTO objects containing segment information
   */
  @Query("""
        SELECT new com.event.processing.notifier.domain.dto.SegmentDTO(
            seg.id, seg.name, subSeg.subscriberId
        )
        FROM SubscriberSegment subSeg
        JOIN Segment seg ON subSeg.segmentId = seg.id
        WHERE subSeg.subscriberId IN :subscriberIds
      """)
  List<SegmentDTO> fetchSegments(Set<String> subscriberIds);
}

File: notifier-go/domain/repository/segment_repository.go
package repository

import (
	"context"
	"database/sql"
	"fmt"

	"github.com/lib/pq"
)

// SegmentRow holds segment data joined with subscriber association
type SegmentRow struct {
	ID           string
	Name         string
	SubscriberID string
}

// SegmentRepository defines the interface for segment data access
type SegmentRepository interface {
	FetchSegments(ctx context.Context, subscriberIDs []string) ([]SegmentRow, error)
}

// segmentRepositoryImpl implements SegmentRepository
type segmentRepositoryImpl struct {
	db *sql.DB
}

// NewSegmentRepository creates a new SegmentRepository
func NewSegmentRepository(db *sql.DB) SegmentRepository {
	return &segmentRepositoryImpl{db: db}
}

// FetchSegments fetches segments associated with the given subscriber IDs
func (r *segmentRepositoryImpl) FetchSegments(ctx context.Context, subscriberIDs []string) ([]SegmentRow, error) {
	if len(subscriberIDs) == 0 {
		return nil, nil
	}

	query := `
		SELECT seg.id, seg.name, sub_seg.subscriber_id
		FROM subscriber_segment sub_seg
		JOIN segment seg ON sub_seg.segment_id = seg.id
		WHERE sub_seg.subscriber_id = ANY($1)
	`

	rows, err := r.db.QueryContext(ctx, query, pq.Array(subscriberIDs))
	if err != nil {
		return nil, fmt.Errorf("failed to query segments: %w", err)
	}
	defer rows.Close()

	var results []SegmentRow
	for rows.Next() {
		var row SegmentRow
		if err := rows.Scan(&row.ID, &row.Name, &row.SubscriberID); err != nil {
			return nil, fmt.Errorf("failed to scan segment row: %w", err)
		}
		results = append(results, row)
	}

	return results, rows.Err()
}
